Enhance project documentation

gszasz opened the pull request:
The `README.md` documentation file need several typo correction and
reorganization of the sections.

- [x] Fix the embarrassing typo in the heading
- [x] Replace old-style Markdown headers with a modern style
- [x] Reorganize sections
- [x] Add *Installation* section
- [x] Add example plot picture
- [x] Add data file example
- [x] Fix issue #1
- [x] Fix *Installation* section according to @sars1492 comment
- [x] Perform minor fixes to make documentation file more consistent

Comments and docstrings and the structure of the code also need be improved.

- [x] Encapsulate initial code into `main()` function
- [x] Improve help `description`, add add *argparse* `epilog`
- [x] Add `__version__` attribute to the module
- [x] Replace module docstring by something more meaningful
- [x] Add constant, class and method docstrings
- [x] Add more comments
- [x] Fix issue #3

gszasz committed:
Refactor README.md file and correct typos

The README.md documentation needed several typo correction and
reorganization of the sections.

 - Replace old-style Markdown headers with a modern style
 - Added picture of a plot
 - Added "Installation" section
 - Added data file example

gszasz committed:
Remove quotes around strings in YAML file

The string values, even those containing spaces, do not have to be
enclosed in quotes of double quotes in YAML syntax.  Although it is good
practice to use these quotes in YAML files anyway, in scope of this
project they are just adding unnecessary complexity.

gszasz committed:
Reformat list of metadata keys to a table

I've decided to use GitHub flavored markdown table feature to make list
of the metadata keys in README.md more readable.

gszasz committed:
Extend metadata key table

I reformatted metadata key table and extended key descriptions.

gszasz committed:
Move example files to 'examples' directory

The example files has to be separated from the files in current
directory to prevent file name collisions in the future.

gszasz committed:
Fix Issue #1

sars1492 commented:
The commands in the **Installation** section of the `README.md` should contain `sudo`, because most users are just copy-pasting the commands from the documentation.

**Suggested fix:**

```
$ sudo dnf install -y python-matplotlib python-yaml
```

```
$ sudo yum install -y python-matplotlib python-yaml
```

```
$ sudo pip install matplotlib yaml
```

gszasz commented:
Thanks @sars1492. I've already included your suggestion it into the pull request.

gszasz committed:
Switch to Liberation Sans font

Originally I was keeping Liberation Serif font because it will just look
better in the printed form; however, it seems that even Springer
recommends Sans Serif fonts for plots and diagrams in order to maintain
text readability when the image is being scaled down.  Therefore I am
giving up and switching to Liberation Sans font, the original default.
Although I still think that Liberation Serif looks nicer in the printed
form, I have to admit that example plot in 'README.md' looks far more
cool now.

gszasz committed:
Fix Installation section of README.md

The commands in the Installation section of the README.md should contain
sudo, because most users are just copy-pasting the commands from the
documentation.

gszasz committed:
Make documentation file more consistent

The allover text of the README.md needed some finishing touches to make
it more thorough and consistent.

 - Rephrase introduction section
 - Extend 'Installation' section
 - Extend 'Usage examples' section
 - Rephrase and correct 'Metadata file' section
 - Swap 'List of valid metadata keys with 'Example metadata file'
 - Fix minor typographic issues in the 'License' section

gszasz committed:
Fix minor markdown formatting issues

gszasz committed:
Alter plot dimensions

I slightly altered the plot dimensions to make figures fit single page
in the LaTeX formatted document.  I also capitalized axis labels in the
default 'plot-metadata.yaml' that seems to be right way to do it.  These
changes should be also reflected in the example plot.

gszasz committed:
Fix miscapitalization in README.md

The 'tab completition' expression is written with lowercase 't'.

gszasz committed:
Encapsulate initial code into main() function

Since local variables are allocated much faster than global variables,
it is a good practice to encapsulate whole initial code into the main()
function.  I also added proper docstring and comments into the code of
the main() function.

gszasz committed:
Define argparse 'description' via string variable

The module docstring serves different purpose and thus it is not a good
idea to populate argparse object 'description' by '__doc__' contents.

gszasz committed:
Refactor PlotMetadataFactory class

Term 'factory' usually refer to a method that is an abstraction of
constructor.  Factory instantiates complex object structure containing
different object types.  The PlotMetadataFactory is more a wrapper class
for the metadata file. It just parses the YAML file and loads metadata
into a list of PlotMetadata objects.

I've decided to rename the class into MetadataFileParser, that is more
intuitive.  I've also always loading all contents of the YAML file and
plot selection part was moved into the select_plots() method.  This
makes code much more readable.  Since I need dictionary only for the
selection, select_plots() method returns list of dictionary values.
Therefore, plot_metadata_dict variable in main() was renamed to
plot_metadata_list.

gszasz committed:
Improve help description and add epilog

I improved help description and epilog to the point that the output of
the command line help can be also used as the module description and
also in 'Synopsis' section of 'README.md'.

gszasz committed:
Add __version__ attribute to the module

Since the code already had a pre-release, it would be fair to specify
current version in the module itself.

gszasz committed:
Refactor 'plot_file' variable to 'output_file'

The variable name 'plot_file' is confusing, because output images are
usually referred in the documentation as 'output files'.

gszasz committed:
Replace 'data sets' with 'datasets'

The language has to be consistent over all documentation.

gszasz committed:
Add constant, class and method docstrings

Added docstrings to every constant, class and method.  The module now
complies PEP 257 guidelines.

gszasz committed:
Add more comments

I needed to reorganize and comment render() method that was originally
copy-pasted from the matplotlib bar chart example and it was quite
incomprehensible.  I also fixed minor bug--Y-axis tick label size was
hardcoded to 11pt, now it can be properly set via YAML file.

gszasz committed:
Fix issue #3

Meanwhile I've figured out the cause of the issue #3 during reorganizing
and commenting of the render() method.  There was single forgotten
debugging line printing out the number of plot datasets.

gszasz commented:
@sars1492 The pull request is ready for review.
